README.md
# SilverStripe in a docker container

SilverStripe is an open-source Content Management System (CMS) and a PHP framework for creating and maintaining websites.

This is a base image for easily running SilverStripe without having to install it first, which is a quite time-consuming process when starting from scratch. 

This image includes:
 - nginx v1.11.2
 - php v5.6.22
 - mysql client v5.5.49
 - composer v1.2-dev
 - zend engine v2.6.0
 - zend opcache v7.0.6-dev

And SilverStripe itself of course! You specify SilverStripe version in the tag of the image: e.g. `gambitlabs/silverstripe:3.1` contains the latest patch release for `3.1` (at the time of writing: `3.1.19`)

_Note: It isn't of production quality (yet!), but we're getting close._

### Dockerize any SilverStripe site! (Express deployment)

If you download this repository, you're able to convert any SilverStripe project to a docker image (at least, we hope so)

```console
$ # Assume you are in the home directory, and have one SilverStripe project called my-project.com
$ pwd
/home/user-1
$ ls
my-project.com

$ # Here you have your SilverStripe project
$ ls my-project.com
assets cms framework your-module some-extension mysite themes _ss_environment.php index.php

$ # Clone this repo
$ git clone https://github.com/gambit-labs/silverstripe

$ # Now we have two folders
$ ls
my-project.com silverstripe
$ cd silverstripe

$ # Deploy the site locally using docker.
$ # SITE_DIR = the directory where the source is
$ # SITE_REPO = the name the docker image should have. In the form docker-user/repository
$ # SITE_VERSION = the version of the site you have.
$ # SS_VERSION = the version of SilverStripe you want to use. Available: [3.0, 3.1, 3.2, 3.3, 3.4, 4.0.0-alpha1]
$ # There are more releases of the Gambit SilverStripe image as well, you may check out Docker Hub for more info
$ make deploy-site-locally SITE_DIR=/home/user-1/my-project.com SITE_REPO=some-docker-hub-id/my-project SITE_VERSION=1.0 SS_VERSION=3.3

$ # Now, check out http://localhost to browse the site
```

### Wait, what? (Manual steps and explanation)

How did that work? Well, since it's dockerized, all setup logic and complexities have been put inside the image, and left are just a few customization parameters. 

No more complex, manual SilverStripe installs!

#### Build an Docker image from your source

The only thing you need to do, is to add a `Dockerfile` to the root of your project, and run `docker build`

```console
$ pwd
/home/user-1/my-project.com

$ # Create a Dockerfile with just one statement, FROM.
$ echo "FROM gambitlabs/silverstripe:3.3" > Dockerfile

$ # And build the image with your site in it
$ docker build -t some-docker-hub-id/my-project:1.0 .
Sending build context to Docker daemon 588.3 kB
Step 1 : FROM gambitlabs/silverstripe:3.3
$ # Executing 1 build trigger...
Step 1 : ADD . ${SOURCE_DIR}
 ---> 27e0d72e4eff
Removing intermediate container ee0ca15ff66d
Successfully built 27e0d72e4eff
```

Here's really some `ONBUILD` magic. It copies the current directory (your site) into a source directory the base image `gambitlabs/silverstripe` serves from. See the Dockerfile in this project and read the Docker documentation about `ONBUILD` for more information.

Note that `cms`, `framework` and `_ss_environment.php` (and `reports`, `siteconfig` when using SilverStripe 3.2 or above) are _NOT_ used in the container.
If they exist, they're overridden by the official SilverStripe release, and this is what it makes it so easy to switch versions, or even run the same site with 5 different SilverStripe versions on the same host, in different Docker containers.

Let's run the website!

```console
$ # This command is a bit complex. First, the content in $() is executed, i.e. docker starts a MariaDB container.
$ # From that run command, the container id is returned. We're using the container id to link the MariaDB container to the SilverStripe container.
$ # From the SilverStripe's point of view, the MariaDB database is available at host db, on port 3306.
$ # We're doing all this without affecting the host at all.

$ # Run MariaDB and SilverStripe. The version may be customized.
$ docker run -d -p 8080:80 \
	--link $(docker run -d -e MYSQL_ROOT_PASSWORD=root mariadb):db \
	-e SS_DATABASE_SERVER=db \
	some-docker-hub-id/my-project:1.0

$ # Now, check out http://localhost:8080 to browse the site
```

The command above, `make deploy-site-locally` did all of this for you, automatically. 
But it's better to learn how it works and set it up manually, in order to understand it fully.

Also note that we may spin up as many containers (=sites) as we want with Docker.
The only thing that should change is the exposed port to the host (the first argument to `-p`)

### Options/customization

You may customize the settings in the `_ss_environment.php` file by passing the arguments as `-e` parameters to `docker run`.

Here are the options and their default values:

SS_* options:
 - `SS_ENVIRONMENT_TYPE=dev`
 - `SS_DATABASE_SERVER=127.0.0.1`
 - `SS_DATABASE_PORT=3306`
 - `SS_DATABASE_USERNAME=root`
 - `SS_DATABASE_PASSWORD=root`
 - `SS_DEFAULT_ADMIN_USERNAME=admin`
 - `SS_DEFAULT_ADMIN_PASSWORD=admin`
 - `SS_DEFAULT_ADMIN_PASSWORD=admin`
 - `SS_ERROR_LOG=silverstripe.errlog`

NGINX_* options:
 - `NGINX_DOMAIN_NAME=localhost`: The domains nginx should serve. Can be a string with multiple domain names. Example: "yyy.com www.yyy.com"
 - `NGINX_LISTEN_PORT=80`: The HTTP port nginx should serve on
 - `NGINX_LISTEN_HTTPS_PORT=443`: The HTTPS port nginx should serve on
 - `NGINX_ENABLE_HTTPS=0`: If HTTPS should be enabled
 - `NGINX_ENABLE_HTTP2=0`: If HTTP/2 should be enabled. Requires HTTPS.
 - `NGINX_WORKER_PROCESSES=1`: The number of worker processed that should be spawned for handling requests.
 - `NGINX_WORKER_CONNECTIONS=1024`: The maximum amount of requests a worker can handle.

PHP_* options:
 - `PHP_SERVER=localhost`: Which server nginx should pass PHP files to.
 - `PHP_MAX_EXECUTION_TIME=300`: How many seconds a PHP request may take.
 - `PHP_MAX_UPLOAD_SIZE=32`: How many megabytes an user is allowed to upload to the server.
 - `PHP_TIMEZONE="Europe/Helsinki"`: The timezone that should be set in php.ini

### Dev mode

By using the method above, the source files of your site aren't modified in any way, as they are copied into an image.

But what if you want to be able to change the source (develop) at the same time you're serving the site?
Enter the development mode!

It's easy, just set `DEV_MODE=1` in the container and mount a path, e.g. `$(pwd)/live` to `/live` in the container. This method won't touch the repo you're working in at all.

You may then edit the source in `$(pwd)/live`, and the changes will be reflected on the site.

```console
$ pwd
/home/user-1/my-project.com
$ ls
assets cms framework your-module some-extension mysite themes _ss_environment.php index.php

$ # Run the image in development mode. The version/port may easily be customized.
$ docker run -d -p 80:80 \
	--link $(docker run -d -e MYSQL_ROOT_PASSWORD=root mariadb):db \
	-e SS_DATABASE_SERVER=db \
	-e DEV_MODE=1 -v $(pwd)/live:/live \
	-v $(pwd):/source \
	gambitlabs/silverstripe:3.3

$ # Notice the "live" folder. The files that are served in the docker container are now there, and you may edit them as you want.
$ # Without touching the source-controlled files in this directory. Perfect!
$ ls
assets cms framework live mysite some-extension themes your-module _ss_environment.php index.php
```

### Readwrite development mode

TODO

## HTTPS and HTTP/2

TODO

### Security

TODO

## Overriding/extending the image config

TODO

### Patching

TODO

## Deploy to Kubernetes!

By using the Makefile that is included in this repo, it's very easy to deploy your newly-built site to Kubernetes for even more scale!

```console
$ # Deploy the site to a Kubernetes cluster.
$ # SITE_DIR = the directory where the source is
$ # SITE_REPO = the name the docker image should have. In the form docker-user/repository
$ # SITE_VERSION = the version of the site you have.
$ # SS_VERSION = the version of SilverStripe you want to use. Available: [3.0, 3.1, 3.2, 3.3, 3.4]
$ # There are more releases of the Gambit SilverStripe image as well, you may check out Docker Hub for more info
$ make deploy-site-k8s SITE_DIR=/home/user-1/my-project.com SITE_REPO=some-docker-hub-id/my-project SITE_VERSION=1.0 SS_VERSION=3.3

$ kubectl get svc my-project -o template --template {{.spec.clusterIP}}
10.0.0.x

$ # Now you may browse your site on http://10.0.0.x

$ # To turn it down, run:
$ make deploy-site-k8s SITE_REPO=some-docker-hub-id/my-project SITE_VERSION=1.0
```

## Build the SilverStripe base image

```console
$ # Variables:
$ # IMAGE_AUTHOR=gambitlabs, the Docker Hub user the images should be pushed to
$ # IMAGE_NAME=silverstripe, the image name
$ # IMAGE_REVISION=1, the revision of the Dockerfile
$ # VERSIONS="4.0.0-alpha1 3.4.0 3.3.2 3.2.4 3.1.19 3.0.14", the versions of SilverStripe we want to build
$ ./build-images.sh
```

and if you want to push at the same time, add `PUSH=1`:

```console
$ PUSH=1 ./build-images.sh
```

The above command will build the SilverStripe base image for all latest 3.x minor releases.

### License

MIT


CHANGELOG.md
## Changelog

v0.3.0
 - HTTPS support! Just mount a `.crt` and a `.key` file, enable it and you're done
 - HTTP 2.0 support! If you have HTTPS enabled, you may use the much faster HTTP standard.
 - Upgraded to nginx 1.11.2 from 1.6.2
 - Securer than before:
   - A lot of new options added to `/etc/nginx/ssl.conf`
   - It's possible to add a file called `dhparam.pem` to `/certs`. Then nginx will use that for encrypting messages before the TLS handshake takes place.
   - Now protected against SS-2015-013
   - Only index.php, install.php, framework/main.php are allowed to execute on the server
   - `install.php` is removed when `mysite/_config.php` is configured properly in a project.
 - Reduced image size with ~135 MB
 - Now it's possible to patch the official SilverStripe repositories just by dropping a file in `_patches/3.4.0` for instance if you want to patch 3.4.0
 - Switched to the MySQL native driver
 - All unnecessary SilverStripe files (like web.config) are removed
 - Allows overrides in `/etc/nginx/silverstripe.d/*.conf`
 - It's now possible to use another PHP server than the built-in one via the `PHP_SERVER` variable.
 - Includes the latest SilverStripe release: `4.0.0-alpha1`
 - gzip compression enabled by default.
 - Now possible to override the nginx config per-repo, by including `*.nginxconf` files in the root of the source code.
 - Added a new development mode: Readwrite mode.
   - Modifies source-controlled repo and let's you test things live with ease
 - Lots of bugfixes
 - Documentation improvements
 - Performance variables:
   - `NGINX_WORKER_PROCESSES`: How many worker processes nginx should have. Default: 1
   - `NGINX_WORKER_CONNECTIONS`: How many connections one process can handle at the same time: Default: 1024
   - `PHP_MAX_EXECUTION_TIME`: Maximum amount of time PHP is allowed to execute. Default: 300 (seconds)
   - `PHP_MAX_UPLOAD_SIZE`: Maximum upload size in megabytes. Default: 32

v0.2.0
 - Initial release on Github; all basic functionality is there.


build-images.sh
#!/bin/bash

# The IMAGE_AUTHOR field is overridable, if you specify your own Docker Hub name, the script will push the images to your account
IMAGE_AUTHOR=${IMAGE_AUTHOR:-gambitlabs}
IMAGE_NAME=${IMAGE_NAME:-silverstripe}
IMAGE_REVISION=${IMAGE_REVISION:-3}
IMAGE_REPO=${IMAGE_AUTHOR}/${IMAGE_NAME}

# Flag if the script should push the images
PUSH=${PUSH:-0}

# The latest SilverStripe versions for all minor releases
LATEST_MINOR_SS_VERSIONS=(

	# Experimental release:
	4.0.0-alpha1

	# Stable release:
	3.4.0

	# Deprecated releases:
	3.3.2
	3.2.4
	3.1.19
	3.0.14
)
VERSIONS=${VERSIONS:-${LATEST_MINOR_SS_VERSIONS[@]}}

# Login on beforehand if we should push
if [[ ${PUSH} == 1 ]]; then
	echo "Logging into Docker Hub. Please specify your password for the user ${IMAGE_AUTHOR}"
	docker login -u ${IMAGE_AUTHOR}
fi

# Loop through the array
for version in ${VERSIONS}; do

	# Build a specific version of SilverStripe, and tag it as the latest one for the minor release 3.x
	docker build --build-arg SILVERSTRIPE_VERSION=${version} -t ${IMAGE_REPO}:${version}-${IMAGE_REVISION} .
	docker tag ${IMAGE_REPO}:${version}-${IMAGE_REVISION} ${IMAGE_REPO}:${version}
	docker tag ${IMAGE_REPO}:${version}-${IMAGE_REVISION} ${IMAGE_REPO}:$(echo ${version} | cut -d. -f1-2)

	# And push the images if we're told to do that
	if [[ ${PUSH} == 1 ]]; then
		docker push ${IMAGE_REPO}:${version}-${IMAGE_REVISION}
		docker push ${IMAGE_REPO}:${version}
		docker push ${IMAGE_REPO}:$(echo ${version} | cut -d. -f1-2)
	fi
done


docker-entrypoint.sh
#!/bin/bash

over-ss-32(){
	SS_MAJOR=$(echo ${SILVERSTRIPE_VERSION} | cut -d. -f1)
	SS_MINOR=$(echo ${SILVERSTRIPE_VERSION} | cut -d. -f2)
	if [[ ${SS_MAJOR} == 3 && $((SS_MINOR < 2)) == 1 ]]; then
		echo "false"
	else
		echo "true"
	fi
}

# SOURCE_DIR and WWW_DIR comes from an ENV variable in the Dockerfile
# Copy everything except cms, framework and _ss_environment.php from ${SOURCE_DIR} to ${WWW_DIR}
if [[ $(ls ${SOURCE_DIR}) != "" ]]; then

	# Also exclude siteconfig and reports when version is over 3.2
	if [[ $(over-ss-32) == "false" ]]; then
		(cd ${SOURCE_DIR} && cp -R $(ls ${SOURCE_DIR} | \
			grep -v 'framework' | \
			grep -v 'cms' | \
			grep -v '_ss_environment.php') ${WWW_DIR})
	else
		(cd ${SOURCE_DIR} && cp -R $(ls ${SOURCE_DIR} | \
			grep -v 'framework' | \
			grep -v 'cms' | \
			grep -v 'reports' | \
			grep -v 'siteconfig' | \
			grep -v '_ss_environment.php') ${WWW_DIR})
	fi
fi

# dev mode exposes all web data at the /live mount.
# to use it, start the docker container normally, but set DEV_MODE=1 to enable it and mount the host dir you want to see the live content to /live
# Example: Expose the web content at $(pwd)/live, where $(pwd) is the SilverStripe project root: 
# docker run -d -e DEV_MODE=1 -v $(pwd):/source -v $(pwd)/live:/live {image-name}
DEV_MODE=${DEV_MODE:-0}
DEV_DIR=${DEV_DIR:-/live}
if [[ ${DEV_MODE} == 1 ]]; then
	(cd ${WWW_DIR} && cp -R ${WWW_DIR}/* ${DEV_DIR})
	rm -r ${WWW_DIR}
	ln -s ${DEV_DIR} ${WWW_DIR}
fi

# readwrite mode copies the official source into the repo you're working on
# the site is hosted from the directory on the host
# WARNING: This will modify your current work directory
# Example: docker run -d -e RW_MODE=1 -v $(pwd):/source {image-name}
RW_MODE=${RW_MODE:-0}
if [[ ${RW_MODE} == 1 && ${DEV_MODE} == 0 ]]; then
	rm -rf ${SOURCE_DIR}/cms ${SOURCE_DIR}/framework ${SOURCE_DIR}/_ss_environment.php

	if [[ $(over-ss-32) == "true" ]]; then
		rm -rf ${SOURCE_DIR}/reports ${SOURCE_DIR}/siteconfig
	fi

	# Do not override existing files with the same name
	cp -r --no-clobber ${WWW_DIR}/* ${SOURCE_DIR}
	rm -r ${WWW_DIR}
	ln -s ${SOURCE_DIR} ${WWW_DIR}
fi

# Traverse SilverStipe patches, and apply them
if [[ -d ${WWW_DIR}/_patches/${SILVERSTRIPE_VERSION} ]]; then
	cd ${WWW_DIR}

	for file in ${WWW_DIR}/_patches/${SILVERSTRIPE_VERSION}/*.patch; do
		echo "Patching SilverStripe with file: $file"
		patch -p1 < $file
	done
fi

# Runtime configuration options
SS_ENVIRONMENT_TYPE=${SS_ENVIRONMENT_TYPE:-dev}
SS_DATABASE_SERVER=${SS_DATABASE_SERVER:-127.0.0.1}
SS_DATABASE_PORT=${SS_DATABASE_PORT:-3306}
SS_DATABASE_USERNAME=${SS_DATABASE_USERNAME:-root}
SS_DATABASE_PASSWORD=${SS_DATABASE_PASSWORD:-root}
SS_DEFAULT_ADMIN_USERNAME=${SS_DEFAULT_ADMIN_USERNAME:-admin}
SS_DEFAULT_ADMIN_PASSWORD=${SS_DEFAULT_ADMIN_PASSWORD:-admin}
SS_ERROR_LOG=${SS_ERROR_LOG:-silverstripe.errlog}

NGINX_DOMAIN_NAME=${NGINX_DOMAIN_NAME:-localhost}
NGINX_LISTEN_PORT=${NGINX_LISTEN_PORT:-80}
NGINX_LISTEN_HTTPS_PORT=${NGINX_LISTEN_HTTPS_PORT:-443}
NGINX_ENABLE_HTTPS=${NGINX_ENABLE_HTTPS:-0}
NGINX_ENABLE_HTTP2=${NGINX_ENABLE_HTTP2:-0}

NGINX_WORKER_PROCESSES=${NGINX_WORKER_PROCESSES:-1}
NGINX_WORKER_CONNECTIONS=${NGINX_WORKER_CONNECTIONS:-1024}

PHP_MAX_EXECUTION_TIME=${PHP_MAX_EXECUTION_TIME:-300}
PHP_MAX_UPLOAD_SIZE=${PHP_MAX_UPLOAD_SIZE:-32}

PHP_TIMEZONE=${PHP_TIMEZONE:-"Europe/Helsinki"}

PHP_SERVER=${PHP_SERVER:-localhost}
MAIL_SERVER=${MAIL_SERVER:-localhost}

# If HTTP/2 is used, HTTPS must also be used
if [[ ${NGINX_ENABLE_HTTP2} == 1 ]]; then
	NGINX_ENABLE_HTTPS=1
fi

cat > ${WWW_DIR}/_ss_environment.php <<EOF
<?php
ini_set('date.timezone', '${PHP_TIMEZONE}');
define('SS_ENVIRONMENT_TYPE', '${SS_ENVIRONMENT_TYPE}');
define('SS_DATABASE_SERVER', '${SS_DATABASE_SERVER}');
define('SS_DATABASE_PORT', '${SS_DATABASE_PORT}');
define('SS_DATABASE_USERNAME', '${SS_DATABASE_USERNAME}');
define('SS_DATABASE_PASSWORD', '${SS_DATABASE_PASSWORD}');
define('SS_DEFAULT_ADMIN_USERNAME', '${SS_DEFAULT_ADMIN_USERNAME}');
define('SS_DEFAULT_ADMIN_PASSWORD', '${SS_DEFAULT_ADMIN_PASSWORD}');
define('SS_ERROR_LOG', '${SS_ERROR_LOG}');
global \$_FILE_TO_URL_MAPPING;
\$_FILE_TO_URL_MAPPING['${WWW_DIR}'] = 'http://${NGINX_DOMAIN_NAME}';
EOF

# Replace dynamic values in the default web site config
sed -e "s|NGINX_WORKER_PROCESSES|${NGINX_WORKER_PROCESSES}|g" -i /etc/nginx/nginx.conf
sed -e "s|NGINX_WORKER_CONNECTIONS|${NGINX_WORKER_CONNECTIONS}|g" -i /etc/nginx/nginx.conf
sed -e "s|PHP_MAX_UPLOAD_SIZE|${PHP_MAX_UPLOAD_SIZE}|g" -i /etc/nginx/nginx.conf

sed -e "s|PHP_MAX_EXECUTION_TIME|${PHP_MAX_EXECUTION_TIME}|g" -i /etc/nginx/php.conf
sed -e "s|PHP_SERVER|${PHP_SERVER}|g" -i /etc/nginx/php.conf

sed -e "s|max_execution_time = 30|max_execution_time = ${PHP_MAX_EXECUTION_TIME}|g" -i /etc/php5/fpm/php.ini
sed -e "s|upload_max_filesize = 2M|upload_max_filesize = ${PHP_MAX_UPLOAD_SIZE}M|g" -i /etc/php5/fpm/php.ini
sed -e "s|post_max_size = 3M|post_max_size = $((PHP_MAX_UPLOAD_SIZE+1))M|g" -i /etc/php5/fpm/php.ini

# Require those two files
# docker run -d -e NGINX_ENABLE_HTTPS=1 -v $(pwd)/certs:/certs {image_name}
if [[ ${NGINX_ENABLE_HTTPS} == 1 && (! -f ${CERT_DIR}/site.crt || ! -f ${CERT_DIR}/site.key) ]]; then

	echo "Fatal error: Tried to start in https mode but ${CERT_DIR}/site.crt or ${CERT_DIR}/site.key does not exist."
	echo "Those two files are required in order to enable https."
	echo "Exiting..."
	exit 1
fi

# If we've enabled https, an optional dhparam.pem file may be specified for added encryption
# If there is no such file, remove the statement from the config
# Generate with this command: openssl dhparam -out /etc/nginx/ssl/dhparam.pem 2048
if [[ ${NGINX_ENABLE_HTTPS} == 1 && ! -f ${CERT_DIR}/dhparam.pem ]]; then
	sed -e "/ssl_dhparam/d" -i /etc/nginx/sites-available/default-https
fi

sed -e "s|SS_ROOT_DIR|${WWW_DIR}|g" -i /etc/nginx/sites-available/default-http /etc/nginx/sites-available/default-https
sed -e "s|NGINX_DOMAIN_NAME|${NGINX_DOMAIN_NAME}|g" -i /etc/nginx/sites-available/default-http /etc/nginx/sites-available/default-https
sed -e "s|NGINX_LISTEN_PORT|${NGINX_LISTEN_PORT}|g" -i /etc/nginx/sites-available/default-http /etc/nginx/sites-available/default-https

sed -e "s|NGINX_LISTEN_HTTPS_PORT|${NGINX_LISTEN_HTTPS_PORT}|g" -i /etc/nginx/sites-available/default-https
sed -e "s|CERT_DIR|${CERT_DIR}|g" -i /etc/nginx/sites-available/default-https

# Set the http2 directive
if [[ ${NGINX_ENABLE_HTTP2} == 1 ]]; then
	sed -e "s|USE_HTTP2|http2|g" -i /etc/nginx/sites-available/default-https
else
	sed -e "s|USE_HTTP2||g" -i /etc/nginx/sites-available/default-https
fi

# If https is disabled, remove the nginx config for HTTPS
mkdir -p /etc/nginx/sites-enabled
if [[ ${NGINX_ENABLE_HTTPS} == 1 ]]; then
	ln -s /etc/nginx/sites-available/default-https /etc/nginx/sites-enabled/default
else
	ln -s /etc/nginx/sites-available/default-http /etc/nginx/sites-enabled/default
fi

# Remove install.php where the $database variable is set in _config.php, i.e. remove install.php from all real projects
if [[ ! -z $(grep "global \$database" ${WWW_DIR}/mysite/_config.php) && -z $(grep "$database = ''" ${WWW_DIR}/mysite/_config.php) ]]; then
	rm -f ${WWW_DIR}/install.php
fi

# Make the user and group www-data own the content. nginx is using that user for displaying content 
chown -R www-data:www-data ${WWW_DIR}

# Only start PHP if we're listening on it here
if [[ ${PHP_SERVER} == "localhost" ]]; then

	# Start the FastCGI server
	exec php5-fpm &
fi

# Start a socat forwarder process if the mail forwarder is present somewhere else
if [[ ${MAIL_SERVER} != "localhost" ]]; then
	exec socat -ls TCP4-LISTEN:25,fork,reuseaddr TCP4:${MAIL_SERVER}:25 &
fi

# Start the nginx webserver in foreground mode. The docker container lifecycle will be tied to nginx.
exec nginx -g "daemon off;"
